### PR TITLE
Support class based commands

### PR DESCRIPTION
**Pull request by gio3k**
Now imports (and initialises) command classes with the same name as the file

Doesn't break support for existing command objects either, they're still imported

### CODE REVIEW

File: src/handlers/CommandHandler.ts [hunk 1]
```diff
@@ -14,20 +14,42 @@ export class CommandHandler {
...
+        return Promise.all(files.map(async file => {
+            const script = file.slice(0, -3) // remove .ts from filename for script name 
+            const commands = await import(join(commandsDir, file))
+
+            if (!commands)
+                return;
+
+            // Check for class type command
+            if (typeof commands[script] === 'function') {
+                const instance = new commands[script]()
```
Comment (giodotblue) [inline src/handlers/CommandHandler.ts:30]: this is on purpose I think it looks nicer

**Comment by lleytonvs** [comment]
I like how this cleans up the command-loading code, but I personally don't like the idea of mixing between object-based, and class-based Command(s). If we choose to use class-based Command(s), then it would make loading object-based Command(s) code obsolete/redundant and vice versa, but if we were to mix between the two, then it breaks any uniformity.

**Comment by giodotblue** [comment]
the plan was to only support class-based commands but I didn't want to make the PR too big by converting all the existing commands

**Comment by lleytonvs** [comment]
The current object-based approach accomplishes everything that we need, so code re-factoring aside, what functional benefit does this bring?

**Comment by giodotblue** [comment]
allows extending commands with extra attributes and methods without creating new interfaces / modifying the current one
lets you base commands off of other commands
handle construction of the command object in the CommandHandler rather than in command scripts
cleaner layout of classes over object creation
can _still_ use the Command interface

the reason it's a better option is because there really are no downsides over an object. it just makes the object!

**Comment by lleytonvs** [comment]
> allows extending commands with extra attributes and methods without creating new interfaces / modifying the current one
lets you base commands off of other commands

Where would this be useful?

> the reason it's a better option is because there really are no downsides over an object. it just makes the object!

If there are no functional benefits to using classes, nor any explicit reason to use them, then there is no upside over what we already use.

**Comment by Wolfleader101** [comment]
Declined based on the reasons above.

There is no reason to use class-based commands. Polymorphism just adds more ambuigity.

**Comment by giodotblue** [comment]
Eventually we as a community have to realise that the bot only currently has 3 commands. We have help, ping and resources.

Of course these can be done with a single object. They are _barely_ scratching the surface of what a bot can do. I don't understand why we need to make it harder for ourselves in the future just because of... "no upside over what we already use"?

> Where would this be useful?
>
Any command that could help from another method or attribute. Really. There's no command that can't gain from this (especially when commands become bigger and more complex) and there's no code style that can't gain from this (it's cleaner, nicer to look at and less TypeScript dependent)

Extending commands off of other commands allows for less bloat; we don't need to have extra files and code bloat to just have multiple commands with the same UI, the same backend functionality, the same anything, why is having more functionality a problem?

> If there are no functional benefits to using classes, nor any explicit reason to use them, then there is no upside over what we already use.
>
I actually don't understand. There were 5 reasons above the message you replied to, it's like you didn't see them at all or something?

The points I made before still hold up so I don't need to put them here again. Really... how in any world is this not an easy & useful change?

**Review by Wolfleader101 — COMMENTED** [review]
I think class-based commands shouldn't be an issue

**Review by Wolfleader101 — APPROVED** [review]
I think class-based commands shouldn't be an issue

**Pull request merged**